ggml/src/ggml-common.h

```diff
@@ -418,6 +433,25 @@ typedef struct {
 } block_iq4_xs;
 static_assert(sizeof(block_iq4_xs) == sizeof(ggml_half) + sizeof(uint16_t) + QK_K/64 + QK_K/2, "wrong iq4_xs block size/padding");
 
+// fp8 support
+// - fp8 simple type
+typedef struct { uint8_t bits; } ggml_e5m2_t;
+typedef struct { uint8_t bits; } ggml_e4m3_t;
+typedef struct { uint8_t bits; } ggml_e3m4_t;
+
+// - fp8 with bloc delta => 8.125 bpw
+typedef struct {
+    float d;  // delta
+    uint8_t qs[QK_K];
+} block_e4m3_q;
+static_assert(sizeof(block_e4m3_q) == sizeof(float) + QK_K, "wrong block_e4m3_q block size/padding");
+
+typedef struct {
+    float d;  // delta
```

Comment:
Would it make sense to use `ggml_half d;`?

Reply:
I don't know.
Native FP8 distributed model use FP32 for there scale, and I use it for all my [experiment](https://github.com/Djip007/llamafile/tree/feature/fp8_study?tab=readme-ov-file#fp8-experiment)

I made many dev with mistral-nemo and there BF16 distribution, with it it may be possible that scale can't be encode as FP16. On my zen4 CPU I use the BF16 dot2 for all matmul compute, with it I need to apply the scale on the result after each bloc before accumulate on C. so at this point I need FP32.

So for simplify I use simple float. with that I get with blocs of 256 fp8 a 8.125bpw.
It may be have sense to use BF16 too in some case (simple convert to FP32)

What do you made for other 'd' when you convert from BF16/FP32 model? did we need to catch some to high/low values?

So for now I think: it may make sense, but is it useful.

Reply:
I wonder if this is indicative of a bigger issue here as all our other quants use FP16 scales. This was fine when we started with Llama 1, which used FP16 weights and therefore worked well with FP16 scales. With Llama 3 and other new BF16 models the weights could potentially exceed the limits of `(quantized weight) * (FP16 scale)`. As per https://github.com/ggerganov/llama.cpp/pull/7150 it looks FP16 vs. BF16 for Llama 3 doesn't matter perplexity wise, but in the long run it might be worth looking at at BF16 scales.

If we can make the BF16 to FP32 conversions as fast as GGML_COMPUTE_FP16_TO_FP32 then at least on the x86 CPU side there shouldn't be a problem with using BF16 scales as the default. I'm not familar with our GPU implementations but I imagine there'll be some trouble there as the kernels were written with native FP16 support in mind.

Reply:
To be fare, I don't thing we have weight that are bigger than FP16_MAX. We may only find number that are lower than FP16_MIN (and not 0...). To have impact on scale we need to have (in case of Q8_0) 32 numbers that is lower than FP16_MIN and higher than BF16_MIN in the same "bloc", very unlikely. and if possible we may only take care of that for have minimal impact on final result.

For GPU I don't know for all but on AMD RDNA3 we have the same compute power with FP16 ans BF16. (and the same with int8...)

Reply:
Seems OK to leave as `float`. In smaller block sizes (e.g. 32) it makes a bigger difference to the bitrate when using `ggml_half` vs `float`. But for block size of 256 the difference is small.